File: leetcode/104. Maximum Depth of Binary Tree/maxDepth.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
function TreeNode(val) {
  this.val = val;
  this.left = null;
  this.right = null;
}

/**
 * @param {TreeNode} root
 * @return {number}
 */
const maxDepth = (root) => {
  let ans = 0;

  const recursive = (node, depth) => {
    if (!node) {
      ans = Math.max(ans, depth);
      return;
    }
    recursive(node.left, depth + 1);
    recursive(node.right, depth + 1);
  };

  recursive(root, 0);

  return ans;
};

const root = new TreeNode(3);
root.left = new TreeNode(9);
root.right = new TreeNode(20);
root.right.left = new TreeNode(15);
root.right.right = new TreeNode(7);

console.log(maxDepth(root));

File: leetcode/104. Maximum Depth of Binary Tree/maxDepth.cpp
#include <algorithm>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

/**
 * @brief 16 m 55 s
 * O(n)
 */
class Solution {
 public:
  int maxDepth(TreeNode *root) { return backtrack(0, root); }

  int backtrack(int depth, TreeNode *node) {
    if (node == nullptr) {
      return depth;
    }

    int left = backtrack(depth + 1, node->left);
    int right = backtrack(depth + 1, node->right);

    return max(left, right);
  }
};
